### DISCUSSION TITLE
Ansible feedback and dynamic inventory

### DISCUSSION THREAD

**Discussion by benasse** [shellhub-io/shellhub#3653] [Show and tell]
Hi everyone,

For about 1 month ansible is working well through Shellhub. This is great news! 🎉 

I wanted to share with the community a script that generates dynamic inventory from a namaspace : https://gist.github.com/benasse/0658a773c089f46d51e7b2b37c82cfd4

At the moment the execution time is not optimal, this comes from the fact that multiplexing ssh / Controll Master is not implemented yet.

A [PR for its support](https://github.com/golang/go/issues/32958) is underway on the golang side, and some seem to have implemented it [on their side](https://github.com/trzsz/trzsz-ssh/blob/main/tssh/control.go).

I don't know how complicated it is to add it to Shellhub.

**Comment by heiytor** [comment]
Hi, @benasse,

I've been investigating this issue, and I can confirm that it has not been fully implemented yet. It seems like ShellHub can initiate a multiplexed session, but it fails to manage multiple "non-background" sessions on the same "master" connection.

For instance, when an authentication is done with `ssh heitor@dev.d8-5e-d3-f1-67-2c@localhost`, it works fine initially. However, subsequent attempts result in failure with the message `Shared connection to localhost closed.` Likewise, when authenticated with `ssh -fN heitor@dev.d8-5e-d3-f1-67-2c@localhost`, the subsequent connection reuses the "master" connection, but subsequent calls still fail with the same `Shared connection to localhost closed.`.

I will work on a fix for this issue and keep you updated. Thank you for reporting it.

**Comment by heiytor** [comment]
I opened issue #3656 to track this

**Comment by benasse** [comment]
oh great, I'll try it when the next RC version is released! 🙂

**Comment by benasse** [comment]
Hi @heiytor ,

I just tested version v0.18.0, connections through ansible are now much faster, multiplexed connections seem to work.
Do you think https://github.com/shellhub-io/shellhub/issues/3656 could be closed ?

**Comment by heiytor** [comment]
Hi @benasse

Thanks for the reply! And yes, we can close the issue now.
